lib/compact.js

```diff
@@ -308,40 +313,100 @@ api.compact = ({
+            // container includes @graph and @id
+            // get or create the map object
+            let mapObject;
+            if(itemActiveProperty in rval) {
+              mapObject = rval[itemActiveProperty];
+            } else {
+              rval[itemActiveProperty] = mapObject = {};
+            }
+            // add compactedItem to map, using value of `@id` or a new blank node identifier
+            _addValue(
+              mapObject, (expandedItem['@id'] || issuer.getId(null)), compactedItem,
+              {propertyIsArray: (!options.compactArrays || container.includes('@set'))});
+          } else if(container.includes('@graph') && container.includes('@index') && _isSimpleGraph(expandedItem)) {
+            // container includes @graph and @index and item is a simple graph
+            // get or create the map object
+            let mapObject;
+            if(itemActiveProperty in rval) {
+              mapObject = rval[itemActiveProperty];
+            } else {
+              rval[itemActiveProperty] = mapObject = {};
+            }
+            // add compactedItem to map, using value of `@index` or `@none`
+            _addValue(
+              mapObject, (expandedItem['@index'] || '@none'), compactedItem,
+              {propertyIsArray: (!options.compactArrays || container.includes('@set'))});
+          } else if(container.includes('@graph') && _isSimpleGraph(expandedItem)) {
+            // container includes @graph but not @id or @index and value is a simple graph object
+            // add compact value
+            _addValue(
+              rval, itemActiveProperty, compactedItem,
+              {propertyIsArray: (!options.compactArrays || container.includes('@set'))});
+          } else {
+            // wrap using @graph alias
+            compactedItem = {
+              [api.compactIri({activeCtx, iri: '@graph'})]: compactedItem
+            };
+
+            // include @id from expanded graph, if any
+            if('@id' in expandedItem) {
+              compactedItem[api.compactIri({activeCtx, iri: '@id'})] =
+                expandedItem['@id'];
+            }
 
-        // handle language and index maps
-        if(container.includes('@language') || container.includes('@index')) {
+            // include @index from expanded graph, if any
+            if('@index' in expandedItem) {
+              compactedItem[api.compactIri({activeCtx, iri: '@index'})] =
+                expandedItem['@index'];
+            }
+            _addValue(
+              rval, itemActiveProperty, compactedItem,
+              {propertyIsArray: (!options.compactArrays || container.includes('@set'))});
+          }
+        } else if(container.includes('@language') || container.includes('@index') ||
+          container.includes('@id') || container.includes('@type')) {
+          // handle language and index maps
           // get or create the map object
-          let mapObject;
+          let mapObject, key;
```

Comment:
Declare vars on new lines:

```js
let mapObject;
let key;
```

But the `key` declaration should really be moved down to where it's immediately relevant, so right above its first use.